File: PageRank/pagerank.m
function[R] = pagerank(A, R, beta, epsilon)
%pagerank function returns R, vector of PageRanks
%arguments:
%A - adjusted adjacency matrix
%    adjacency matrix divided by out-degree
%    pre-processing step to take care of dead ends
%    M = M + (1-sum(M))/rows(M)
%    K = diag(sum(M))
%    A = K^-1 * M
%    obviously we don't want to preprocess M

%R - initial ranking vector, usually all 1/N
%beta - dampening factor = teleport probability
%epsilon - convergence threshold

N = size(A)(1);
% E = matrix one 1s, used to broadcast teleportation
E = ones(N);
R_last = R;
delta=epsilon;
count=0;
while (delta >= epsilon)
%while (count <= 4)
    R = (beta * A + (1-beta)/N * E) * R_last;
    dist = R-R_last;
    delta = sqrt(dot(dist,dist));
    R_last = R;
    count=count+1
    R
end

File: PageRank/PageRank.py
import numpy as np

def PageRank(M, beta, epsilon):
    N = M.shape[0]
    # array of 1s used to multiply teleportation probability
    E = np.ones((N,N))
    R = np.zeros(N)
    R[R==0] = 1/float(N)
    R_last = R
    delta = epsilon
    while delta >= epsilon:
        #compute PageRank R
        R = np.dot((beta * M + (1-beta)/N * E ) , R_last)
        # replace lost votes from dead ends
        R = R + (1-np.sum(R))/len(R)
        dist = R - R_last
        delta = np.sqrt(np.dot(dist,dist))
        R_last = R
    return R

# example use of the function
M = np.array([[0,0,1],[0.5,0,0],[0.5,1,0]])
beta = 0.85
epsilon = 0.01
print(PageRank(M,beta,epsilon))
